### ISSUE TITLE
Local file imports

### ISSUE THREAD

**Issue by nandorojo** [val-town/deno-http-worker#15]
Is there a way for the worker file to use local file imports?

I tried to import a simple test file called `./import.ts`. However, it throws this error:

```tsx
node index.js
[deno] error: Uncaught (in promise) TypeError: invalid URL: relative URL with a cannot-be-a-base base
[deno]     at data:text/tsx,import%20testImport%20from%20'.%2Fimport.ts'%0A%20%20%0A%20%20export%20default%20%7B%0A%20%20%20%20%20%20%20%20async%20fetch(req%3A%20Request)%3A%20Promise%3CResponse%3E%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20Response.json(%7B%20ok%3A%20req.url%2C%20testImport%20%7D)%3B%0A%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%7D:1:24
[deno] const mod = await import(importURL);
[deno]             ^
```

### Reproduction

```tsx
import { newDenoHTTPWorker } from 'deno-http-worker'
import fs from 'fs'

const files = {
  './import.ts': `export default 1`,
}

for (const file in files) {
  fs.writeFileSync(file, files[file])
}

const worker = await newDenoHTTPWorker(
  `import testImport from './import.ts'
  
  export default {
        async fetch(req: Request): Promise<Response> {
            return Response.json({ ok: req.url, testImport });
        },
    }`,
  { printOutput: true, runFlags: ['--allow-net'] }
)
```

<details>
<summary>Click me for the full code</summary>

```tsx
import { newDenoHTTPWorker } from 'deno-http-worker'
import fs from 'fs'

const files = {
  './import.ts': `export default 1`,
}

for (const file in files) {
  fs.writeFileSync(file, files[file])
}

const worker = await newDenoHTTPWorker(
  `import testImport from './import.ts'
  
  export default {
        async fetch(req: Request): Promise<Response> {
            return Response.json({ ok: req.url, testImport });
        },
    }`,
  { printOutput: true, runFlags: ['--allow-net'] }
)

const body = await new Promise((resolve, reject) => {
  const req = worker.request('https://hello/world?query=param', {}, (resp) => {
    const body = []
    resp.on('error', reject)
    resp.on('data', (chunk) => {
      body.push(chunk)
    })
    resp.on('end', () => {
      resolve(Buffer.concat(body).toString())
    })
  })
  req.end()
})
console.log(body) // => {"ok":"https://hello/world?query=param"}

worker.terminate()
```

</details>

**Comment by maxmcd** [comment]
@nandorojo thanks for the lovely example and instructions to reproduce. 

I wasn't aware of this limitation, but I guess it makes sense. When you pass the source as a string it is passed as a  urlencoded source (as apposed to an HTTP url or filesystem path) so Deno doesn't know what the relative import should be relative to.

I was able to get it working if you pass the source as a file, like so:

```ts
const files = {
"./import.ts": `export default 1`,
"./handler.ts": `import testImport from './import.ts'

export default {
      async fetch(req: Request): Promise<Response> {
          return Response.json({ ok: req.url, testImport });
      },
  }
`,
};

for (const file in files) {
fs.writeFileSync(file, files[file]);
}

const worker = await newDenoHTTPWorker(
new URL(
  "file:///Users/maxm/go/src/github.com/val-town/deno-http-worker/handler.ts"
),
{
  printOutput: true,
  runFlags: ["--allow-net", "--allow-read=./import.ts"],
}
);

const body = await new Promise((resolve, reject) => {
const req = worker.request(
  "https://hello/world?query=param",
  {},
  (resp) => {
    const body = [];
    resp.on("error", reject);
    resp.on("data", (chunk) => {
      body.push(chunk);
    });
    resp.on("end", () => {
      resolve(Buffer.concat(body).toString());
    });
  }
);
req.end();
});
console.log(body); // => {"ok":"https://hello/world?query=param"}

worker.terminate();
```

You'll need to replace my hardcoded path with the absolute path to the files you're writing.

**Comment by maxmcd** [comment]
Ah, alternatively you can also make the import in your script absolute instead:
```ts
const worker = await newDenoHTTPWorker(
  `import testImport from 'file:///Users/maxm/deno-http-worker/import.ts'

  export default {
        async fetch(req: Request): Promise<Response> {
            return Response.json({ ok: req.url, testImport });
        },
    }`,
  { printOutput: true, runFlags: ["--allow-net", "--allow-read=import.ts"] }
);
```

**Comment by nandorojo** [comment]
Ahh super cool, let me try this now.

**Comment by nandorojo** [comment]
@maxmcd thanks for the fast reply btw, I just tried this out, but it didn't appear to work for me...

For some reason, the first case where you're doing `new URL()` doesn't seem to work for me. 

## Error

I'm getting this...

```sh
reject(Object.assign(new Error("Deno exited before being ready"), {
                                     ^

Error: Deno exited before being ready
    at ChildProcess.<anonymous> (file:///Users/fernandorojo/Developer/q2-2024/drite-backend/app/vendors/deno_cloud/javascript/deno/worker/bun/node_modules/.deno/deno-http-worker@0.0.12/node_modules/deno-http-worker/dist/DenoHTTPWorker.js:92:38)
```


## Reproduction

Here's my repro (tried to make it generalizable).

```
nvm use 19
node index.js
```

In the code below, I'm getting the actual path dynamically, and I've confirmed that the `URL()` is indeed getting the correct path.

```ts
// index.js

import { newDenoHTTPWorker } from 'deno-http-worker'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url) // get the resolved path to the file
const __dirname = path.dirname(__filename) // get the name of the directory

const files = {
  './import.ts': `export default 1`,
  './handler.ts': `import testImport from './import.ts'
  
  export default {
        async fetch(req: Request): Promise<Response> {
            return Response.json({ ok: req.url, testImport });
        },
    }`,
}

for (const file in files) {
  fs.writeFileSync(file, files[file])
}

const filePath = `file://` + path.resolve(__dirname, './handler.ts')

console.log({ filePath })

const worker = await newDenoHTTPWorker(new URL(filePath), {
  printOutput: true,
  runFlags: ['--allow-net'],
})

const body = await new Promise((resolve, reject) => {
  const req = worker.request('https://hello/world?query=param', {}, (resp) => {
    const body = []
    resp.on('error', reject)
    resp.on('data', (chunk) => {
      body.push(chunk)
    })
    resp.on('end', () => {
      resolve(Buffer.concat(body).toString())
    })
  })
  req.end()
})
console.log(body) // => {"ok":"https://hello/world?query=param"}

worker.terminate()
```

**Comment by nandorojo** [comment]
@maxmcd thanks for the fast reply btw, I just tried this out, but it didn't appear to work for me...

For some reason, the first case where you're doing `new URL()` doesn't seem to work. 

## Error

I'm getting this...

```sh
reject(Object.assign(new Error("Deno exited before being ready"), {
                                     ^

Error: Deno exited before being ready
    at ChildProcess.<anonymous> (file:///Users/fernandorojo/Developer/worker/node_modules/.deno/deno-http-worker@0.0.12/node_modules/deno-http-worker/dist/DenoHTTPWorker.js:92:38)
```


## Reproduction

Here's my repro (tried to make it generalizable).

```
nvm use 19
node index.js
```

In the code below, I'm getting the actual path dynamically, and I've confirmed that the `URL()` is indeed getting the correct path.

```ts
// index.js

import { newDenoHTTPWorker } from 'deno-http-worker'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url) // get the resolved path to the file
const __dirname = path.dirname(__filename) // get the name of the directory

const files = {
  './import.ts': `export default 1`,
  './handler.ts': `import testImport from './import.ts'
  
  export default {
        async fetch(req: Request): Promise<Response> {
            return Response.json({ ok: req.url, testImport });
        },
    }`,
}

for (const file in files) {
  fs.writeFileSync(file, files[file])
}

const filePath = `file://` + path.resolve(__dirname, './handler.ts')

console.log({ filePath })

const worker = await newDenoHTTPWorker(new URL(filePath), {
  printOutput: true,
  runFlags: ['--allow-net'],
})

const body = await new Promise((resolve, reject) => {
  const req = worker.request('https://hello/world?query=param', {}, (resp) => {
    const body = []
    resp.on('error', reject)
    resp.on('data', (chunk) => {
      body.push(chunk)
    })
    resp.on('end', () => {
      resolve(Buffer.concat(body).toString())
    })
  })
  req.end()
})
console.log(body) // => {"ok":"https://hello/world?query=param"}

worker.terminate()
```

**Comment by nandorojo** [comment]
Weirdly I'm getting that error on both examples you sent. What's your deno version?

I'm on `1.44.1`.

**Comment by nandorojo** [comment]
Ah shoot, disregard, I got it working. Thanks for your help!

**Comment by maxmcd** [comment]
@nandorojo fyi there is a bug where certain promises will never resolve. you can see a repro in the current open PR. it should be fixed in Deno 1.44.2
